Stratis.Bitcoin/BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using NBitcoin;
+using NBitcoin.Protocol;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.BlockStore
+{
+    public class BlockStore
+    {
+	    private readonly ConcurrentChain concurrentChain;
+	    private readonly BlockRepository blockRepository;
+	    private readonly TaskFactory taskFactory;
+
+	    public BlockStore(ConcurrentChain concurrentChain, BlockRepository blockRepository)
+	    {
+		    this.concurrentChain = concurrentChain;
+		    this.blockRepository = blockRepository;
+			this.taskFactory = new TaskFactory(new CustomThreadPoolTaskScheduler(2, 500, "BlockStore"));
+	    }
+
+	    public Task ProcessGetData(Node node, GetDataPayload getDataPayload)
+	    {
+		    return this.taskFactory.StartNew(async () =>
+		    {
+			    foreach (var item in getDataPayload.Inventory.Where(inv => inv.Type == InventoryType.MSG_BLOCK))
+			    {
+				    var block = await this.blockRepository.GetAsync(item.Hash);
+
+				    if (block != null)
+					    node.SendMessage(new BlockPayload(block));
+			    }
+		    });
+	    }
+
+		public Task ProcessGetBlocks(Node node, GetBlocksPayload getBlocksPayload)
+		{
+			return this.taskFactory.StartNew(() =>
+			{
+				ChainedBlock chainedBlock = null;
+				foreach (var item in getBlocksPayload.BlockLocators.Blocks)
```

Comment:
the BlockLocator is not a list a blocks actually, but a way for the full node to know where the client is forking with the full node's chain. You should use `concurrentChain.FindFork(getBlocksPayload.BlockLocators)`, then sends the blocks from the fork (excluded), to the Tip validated by the consensus loop.